### PR TITLE
Burning Buildings Bugfix

### LINKED ISSUE #56242: Burning building results in spam about terrain t_open_air where furniture is f_rubble
### Describe the bug

 DEBUG    : Setting terrain t_open_air at (127,42,1) where furniture is f_rubble.  Terrain is_open_air
If this is intentional, set the ALLOW_ON_OPEN_AIR flag on the furniture

 FUNCTION : ter_set
 FILE     : D:\a\Cataclysm-DDA\Cataclysm-DDA\src\map.cpp
 LINE     : 1769
 VERSION  : 53f9bd3

This happens a lot when I accidentally set an apartment tower on fire (I intended to set fire to a hedge maze adjacent to it).

### Steps to reproduce

Partial a guess:

1. Set fire to a building higher than one story.
2. Watch spam until you give up and ignore it.

### Expected behavior

No spam when buildings burn and things cave in (a guess is that zombies smashing through buildings ought to be capable of generating similar spam as well).

### Screenshots

Offset screen image of the hedge maze and adjacent apartment tower:

Screenshot (226)

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19043.1586 (21H1)
- Game Version: 53f9bd3 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]
+ "mod" in the form of hacking portal storms to do nothing as the actual mod to do so was rejected.

### Additional context

_No response_

### LINKED ISSUE #60973: Some buildings/objects seem to burn forever (months+)
### Describe the bug

I lit a bed in an adobe walled home on fire near my base a few days ago. Now I keep getting this error message. The fire is not spreading and is contained as a small fire on the bed. I've waited an hour next to the fire and it doesn't happen.

I've seen this message before on a previous character, and the house on fire kept burning for months even when living right next to it. The following error message is spammed anytime I'm within a few tiles of the burning object.

 DEBUG    : Setting terrain t_open_air at (20,126,0) where furniture is f_rubble.  Terrain is_open_air
If this is intentional, set the ALLOW_ON_OPEN_AIR flag on the furniture

 FUNCTION : bool map::ter_set(const tripoint&, const ter_id&, bool)
 FILE     : src/map.cpp
 LINE     : 1975
 VERSION  : b9c5366

### Steps to reproduce

I'm not entirely sure. The house I lit was an adobe walled house. Will include a screencap and debug below.

### Expected behavior

Houses generally burn down within a day or so. 

### Screenshots

[debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/9561079/debug.log)
image

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19043.1889 (21H1)
- Game Version: b9c5366 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    No Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions]
]

### Additional context

_No response_

### LINKED ISSUE #66822: Repeatedly getting "Setting terrain t_hole at X where furniture is f_rubble. Terrain is_open_air."
### Describe the bug

Playing for the first time, exited the gas station where I started, moved to the north and when I started moving to the west, I get this error constantly. 

It was not only at `(10,114,0)`, but when pressing space to continue, the error keeps popping up for different coordinates.

```
 DEBUG    : Setting terrain t_hole at (10,114,0) where furniture is f_rubble.  Terrain is_open_air
If this is intentional, set the ALLOW_ON_OPEN_AIR flag on the furniture

 FUNCTION : bool map::ter_set(const tripoint&, const ter_id&, bool)
 FILE     : src/map.cpp
 LINE     : 2114
 VERSION  : 96c7690
```

Then when reloading the save game again, I get this one:

```
 DEBUG    : error: item condom cannot fit into pocket while loading: item is too heavy

 FUNCTION : void item_contents::combine(const item_contents&, bool, bool, bool)
 FILE     : src/item_contents.cpp
 LINE     : 705
 VERSION  : 96c7690
```

### Attach save file

[Fay-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/12016086/Fay-trimmed.tar.gz)

### Steps to reproduce

Honestly, I don't know whether it's reproducible. I started the game with **everything random**, and as said above, as soon as I left the gas station where I started on and started moving in the world, I got the errors.

### Expected behavior

I'd expect to these errors to not happen in the generated world?

### Screenshots

_No response_

### Versions and configuration

- Windows 10
- v: `96c7690`
- Tiles
- English

### Additional context

- [debug.log](https://github.com/CleverRaven/Cataclysm-DDA/files/12016128/debug.log)
- [options.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/12016122/options.zip)

### LINKED ISSUE #68842: Collapsed tower bugs
### Describe the bug

Various bugs related to the flooring in collapsed tower

1. Generated holes in the floor destroy the objects that occupied the space, leaving the debris "floating" above the hole. This debris will update and fall AFTER performing actions such as climbing down, or placing a stepladder on the tile below, resulting in a smashed stepladder or face. 

2. Holes can generate above the nanofab room, allowing the player to climb directly down into the target room. I'm assuming this is not intended, as doing so can (sometimes?) trigger the hole being inexplicably replaced by a metal floor, sealing you below. #66822 probably related, as described by ANID in the comments. Seems like saving and loading within the reality bubble triggers the metal flooring to spawn as well, which I did not observe myself.

### Attach save file

Not available, will try to set one up if others are unable to verify or reproduce.

### Steps to reproduce

1. Find and visit a collapsed tower, climb down a hole, be smashed by falling debris from above.

2. Not sure what triggers the metal flooring/roof to spawn.

### Expected behavior

1. Debris to fall upon loading into the reality bubble or when holes are generated.

2. Metal floor/roof in place when holes are generated.

### Screenshots

_No response_

### Versions and configuration

OS: Android
OS Version: Manufacturer: OnePlus; Model: HD1900; Release: 12; Incremental: Q.202303230244;
Game Version: 2c8ae07 [64-bit]
Graphics Version: Tiles
Game Language: System language []
Mods loaded: [
  Dark Days Ahead [dda],
  Disable NPC Needs [no_npc_food],
  Slowdown Fungal Growth [no_fungal_growth],
  Bionic Professions [package_bionic_professions]
]

### Additional context

_No response_

### LINKED ISSUE #74014: Burning building results in spam about terrain t_open_air where furniture is f_rubble
### Describe the bug

Previously occurred in #56242 but seems to not be fixed in current experimental

 DEBUG    : Setting terrain t_hole at (102,76,1) where furniture is f_rubble.  Terrain is_open_air
If this is intentional, set the ALLOW_ON_OPEN_AIR flag on the furniture

 FUNCTION : bool map::ter_set(const tripoint&, const ter_id&, bool)
 FILE     : src/map.cpp
 LINE     : 2266
 VERSION  : cdda-experimental-2024-05-22-0114 970efaa

### Attach save file

[Briaroaks-trimmed.tar.gz](https://github.com/CleverRaven/Cataclysm-DDA/files/15410725/Briaroaks-trimmed.tar.gz)

### Steps to reproduce

Either:
load the save
wait a few in game minutes to get the error to pop up

Or:
set a building on fire (might need to be 2 stories)
wait a few in game hours for the error to pop up

### Expected behavior

I expect that a burning building doesn't try to place rubble on open air

### Screenshots

_No response_

### Versions and configuration

- OS: Windows
    - OS Version: 10.0.19045.4291 (22H2)
- Game Version: cdda-experimental-2024-05-22-0114 970efaa [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Portal Storms Ignore NPCs [personal_portal_storms],
    Slowdown Fungal Growth [no_fungal_growth],
    Bionic Professions [package_bionic_professions],
    Magiclysm [magiclysm],
    Alternative Map Key [alt_map_key],
    Mythical Martial Arts [MMA],
    Bionic Slots [cbm_slots]
]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ShnitzelX2**
#### Summary
Bugfixes "bugfixes related to burning buildings"

#### Purpose of change

- Fixes #56242
- Fixes #60973
- Fixes #66822
- Fixes #74014

#### Describe the solution

There were multiple issues, largely caused by this snippet of code in bash_ter_furn():
`if( smash_ter && ter( p )->has_flag( "EMPTY_SPACE" ) && zlevels ) {`
`     tripoint_bub_ms below = p + tripoint_rel_ms_below; `
`    const ter_str_id roof = get_roof( below, params.bash_floor && ter( below ).obj().movecost != 0 );`
`    ter_set( p, roof );`
`}`
When a tile above terrain with the "roof" field is destroyed, the code above adds that roof. It failed to account for that tile already being that roof, so e.g. t_wood_roof above t_door_c could burn forever because it'd get immediately replaced. If there happened to be e.g. f_rubble on the tile, then you'd get the ALLOW_ON_OPEN_AIR debug spam. Example: molotovs on t_wood_roof with t_door_c below versus without for in-game hours:
fire1

This also caused another error (probably unreported): because get_roof() returns t_dirt or t_open_air by default, dirt tiles would propagate upwards when terrain without the "roof" was destroyed, and air tiles would be placed onto f_rubble, also throwing the debug error. Example (before/after fix):
fire2

Finally, fixing all of this also revealed yet another error with explosions: one-tile-deep pits were being automatically filled in by the above snippet, so explosion craters were smaller and steeper than they should have been. Example: 100L bomb (before/after fix):
bomb1

These issues were fixed by:
1) Having roof replacement check for same tile and if the tile below has the WALL flag
2) Fires use bash_floor because they should destroy floors
3) If a floor is destroyed and it has bash_below=true, then also destroy the tile underneath

#### Testing

Burnt down different types of buildings, including multi-story buildings. I didn't see any errors, but report them again if you see them.

#### Additional context

Pinging @PatrikLundell because this is roof work and map.cpp was changed

Terrain roofs need an audit. For example, t_wall is FLAMMABLE, but its "roof" (t_flat_roof) is not. 

Implied, but #66352 did not fix #56242 because it only took into account fire processing of tiles with FLAMMABLE_ASH.

### CODE REVIEW

**Comment by PatrikLundell** [comment]
Yes, it's good not to regenerate a roof after it has been removed. I'm undecided on whether generating (potentially removed) roofs on top of walls without roofs is good, though. I have refrained from adding roofs on walls used as barriers (those surrounding compounds of various sorts) and on all walls of ruined houses (where some walls may retain their roofs, and some don't).

The problem is really one of being able to get over a wall without a roof. Currently you can't get over it, but you can get on top of it using e.g. a step ladder if it has a roof. There's really a need for a mechanism to pass over a roofless wall that isn't a fence, but with more effort and/or tools.

I'm not sure removing the holes that now won't close up from the collapsed tower is desirable, but that's a matter of taste. On the other hand, the collapsed tower doesn't look like one as it never can have had anything above the ground floor (look at buildings collapsed by earthquakes and explosives and you'll see huge amounts of rubble several stories high, typically blocking access to any basements that weren't crushed outright).

**Comment by ShnitzelX2** [comment]
I removed the collapsed tower edit, it was in retrospect out of scope for this PR.

Also, I don't know what's up with the "overmap_terrain_coverage" test failing for Clang12, but I ran that test three times on my branch and three times on master, and they all passed on my end.

**Comment by PatrikLundell** [comment]
The tests have been restarted, so I can't see what went wrong, but a lot of tests are flaky and fail randomly for unrelated reasons. The powers that be prefer to have false negative tests rather than false positive ones (i.e. failures when there's nothing wrong, rather than missing things that should fail), but are at the same time don't placing improving the tests as a particularly high priority.

File: src/map.cpp [hunk 1]
```diff
@@ -4454,29 +4462,31 @@ void map::bash_ter_furn( const tripoint_bub_ms &p, bash_params &params )
...
 
     if( !tent ) {
         spawn_items( p, item_group::items_from( bash->drop_group, calendar::turn ) );
     }
-
-    if( smash_ter && ter( p )->has_flag( "EMPTY_SPACE" ) && zlevels ) {
-        tripoint_bub_ms below = p + tripoint_rel_ms_below;
+    //regenerates roofs for tiles that should be walkable from above
+    if( smash_ter && ter( p )->has_flag( "EMPTY_SPACE" ) && ter( below )->has_flag( "WALL" ) &&
+        zlevels && !set_to_air ) {
```
Comment (ZhilkinSerg) [inline src/map.cpp:4489]:
```suggestion
    if( zlevels && smash_ter && !set_to_air  && ter( p )->has_flag( "EMPTY_SPACE" ) && 
        ter( below )->has_flag( "WALL" ) ) {
```

Comment (ZhilkinSerg) [reply src/map.cpp:4489]: It is faster to check for zlevels first than checking terrain flags.

**Pull request merged**